### PR TITLE
Enhancement/skale 2837 review and test broadcast and complaint

### PR DESCRIPTION
**Pull request by payvint**
Add new type of complaint and clean a little bit

### CODE REVIEW

File: contracts/SkaleDKG.sol [hunk 1]
```diff
@@ -152,23 +157,36 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
-            complaints[groupIndex].nodeToComplaint = toNodeIndex;
-            complaints[groupIndex].fromNodeToComplaint = fromNodeIndex;
-            complaints[groupIndex].startComplaintBlockTimestamp = block.timestamp;
-            emit ComplaintSent(groupIndex, fromNodeIndex, toNodeIndex);
-        } else if (broadcasted && complaints[groupIndex].nodeToComplaint != toNodeIndex) {
-            return;
+            // incorrect data or missing alright
+            if (
+                isEveryoneBroadcasted(groupIndex) &&
+                dkgProcess[groupIndex].startAlrightTimestamp.add(complaintTimelimit) <= block.timestamp
```
Comment (olehnikolaiev) [inline contracts/SkaleDKG.sol]: check alright wasnt sent

**Review by olehnikolaiev — DISMISSED** [review]
good!

File: contracts/SkaleDKG.sol [hunk 2]
```diff
@@ -145,30 +156,44 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
-            complaints[groupIndex].nodeToComplaint = toNodeIndex;
-            complaints[groupIndex].fromNodeToComplaint = fromNodeIndex;
-            complaints[groupIndex].startComplaintBlockTimestamp = block.timestamp;
-            emit ComplaintSent(groupIndex, fromNodeIndex, toNodeIndex);
-        } else if (broadcasted && complaints[groupIndex].nodeToComplaint != toNodeIndex) {
-            return;
+            // incorrect data or missing alright
+            if (
+                isEveryoneBroadcasted(groupIndex) &&
+                startAlrightTimestamp[groupIndex].add(COMPLAINT_TIMELIMIT) <= block.timestamp &&
```
Comment (olehnikolaiev) [inline contracts/SkaleDKG.sol:169]: mb startComplaintTimestamp + comllaint_timelimit?

File: contracts/SkaleDKG.sol [hunk 3]
```diff
@@ -145,30 +156,44 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
+                !isAllDataReceived(groupIndex, toNodeIndex)
+            ) {
+                // missing alright
+                _finalizeSlashing(groupIndex, toNodeIndex);
+            } else {
+                // incorrect data
+                complaints[groupIndex].nodeToComplaint = toNodeIndex;
+                complaints[groupIndex].fromNodeToComplaint = fromNodeIndex;
+                complaints[groupIndex].startComplaintBlockTimestamp = block.timestamp;
```
Comment (olehnikolaiev) [inline contracts/SkaleDKG.sol:178]: for alrightTimestamp we use `now` and for complaintTimestamp - `block.timestamp`. what is the difference?

File: contracts/SkaleDKG.sol [hunk 4]
```diff
@@ -230,14 +255,26 @@ contract SkaleDKG is Permissions, ISkaleDKG {
         return dkgProcess[groupIndex].numberOfCompleted;
     }
 
-    function isChannelOpened(bytes32 groupIndex) external override view returns (bool) {
```
Comment (olehnikolaiev) [inline contracts/SkaleDKG.sol:233]: do we really need to remove this call?

File: contracts/SkaleDKG.sol (hunk 3 above)
Comment (payvint) [reply contracts/SkaleDKG.sol:178]: So there are no difference. That's our code style:D

File: contracts/SkaleDKG.sol [hunk 5]
```diff
@@ -145,30 +156,44 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
-        } else if (broadcasted && complaints[groupIndex].nodeToComplaint != toNodeIndex) {
-            return;
+            // incorrect data or missing alright
+            if (
+                isEveryoneBroadcasted(groupIndex) &&
+                startAlrightTimestamp[groupIndex].add(COMPLAINT_TIMELIMIT) <= block.timestamp &&
+                !isAllDataReceived(groupIndex, toNodeIndex)
+            ) {
+                // missing alright
+                _finalizeSlashing(groupIndex, toNodeIndex);
```
Comment (olehnikolaiev) [inline contracts/SkaleDKG.sol:173]: if some malicious node will not send a broadcast and once broadcast timeout expires send a complaint on one of the good nodes then this good node will be slashed, won't it?

File: contracts/SkaleDKG.sol (hunk 3 above)
Comment (payvint) [reply contracts/SkaleDKG.sol:178]: P.s. I will change `block.timestamp` to `now` everywhere

File: contracts/SkaleDKG.sol (hunk 5 above)
Comment (payvint) [reply contracts/SkaleDKG.sol:173]: This good node should send response. Let me recheck

**Review by yavrsky — APPROVED** [review]

File: contracts/SkaleDKG.sol (hunk 5 above)
Comment (payvint) [reply contracts/SkaleDKG.sol:173]: So good point. And in this case - good node should send response.
But we need to add a check that complaint could send only after broadcast, if node have not broadcasted and time expires, it should be impossible to send complaint.
Does it make sense?

Comment (olehnikolaiev) [reply contracts/SkaleDKG.sol:173]: yes, it would be great to have!

**Review by cstrangedk — APPROVED** [review]

**Review by olehnikolaiev — APPROVED** [review]
great!

**Review by DimaStebaev — APPROVED** [review]

**Pull request merged**